File: packages/firebase_common/lib/src/firestore_documents/host.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'host.dart';

class ReadHost {
  const ReadHost({
    required this.hostId,
    required this.path,
    required this.imageUrl,
    required this.displayName,
    required this.introduction,
    required this.hostTypes,
    required this.urls,
    required this.createdAt,
    required this.updatedAt,
  });

  final String hostId;

  final String path;

  final String imageUrl;

  final String displayName;

  final String introduction;

  final Set<HostType> hostTypes;

  final List<String> urls;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  factory ReadHost._fromJson(Map<String, dynamic> json) {
    return ReadHost(
      hostId: json['hostId'] as String,
      path: json['path'] as String,
      imageUrl: json['imageUrl'] as String? ?? '',
      displayName: json['displayName'] as String? ?? '',
      introduction: json['introduction'] as String? ?? '',
      hostTypes: json['hostTypes'] == null
          ? const <HostType>{}
          : _hostTypesConverter.fromJson(json['hostTypes'] as List<dynamic>?),
      urls:
          (json['urls'] as List<dynamic>?)?.map((e) => e as String).toList() ??
              const <String>[],
      createdAt: (json['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadHost.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadHost._fromJson(<String, dynamic>{
      ...data,
      'hostId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateHost {
  const CreateHost({
    this.imageUrl = '',
    required this.displayName,
    required this.introduction,
    required this.hostTypes,
    this.urls = const <String>[],
  });

  final String imageUrl;
  final String displayName;
  final String introduction;
  final Set<HostType> hostTypes;
  final List<String> urls;

  Map<String, dynamic> toJson() {
    return {
      'imageUrl': imageUrl,
      'displayName': displayName,
      'introduction': introduction,
      'hostTypes': _hostTypesConverter.toJson(hostTypes),
      'urls': urls,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateHost {
  const UpdateHost({
    this.imageUrl,
    this.displayName,
    this.introduction,
    this.hostTypes,
    this.urls,
    this.createdAt,
  });

  final String? imageUrl;
  final String? displayName;
  final String? introduction;
  final Set<HostType>? hostTypes;
  final List<String>? urls;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    return {
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (displayName != null) 'displayName': displayName,
      if (introduction != null) 'introduction': introduction,
      if (hostTypes != null)
        'hostTypes': _hostTypesConverter.toJson(hostTypes!),
      if (urls != null) 'urls': urls,
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteHost {}

/// Provides a reference to the hosts collection for reading.
final readHostCollectionReference =
    FirebaseFirestore.instance.collection('hosts').withConverter<ReadHost>(
          fromFirestore: (ds, _) => ReadHost.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a host document for reading.
DocumentReference<ReadHost> readHostDocumentReference({
  required String hostId,
}) =>
    readHostCollectionReference.doc(hostId);

/// Provides a reference to the hosts collection for creating.
final createHostCollectionReference =
    FirebaseFirestore.instance.collection('hosts').withConverter<CreateHost>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a host document for creating.
DocumentReference<CreateHost> createHostDocumentReference({
  required String hostId,
}) =>
    createHostCollectionReference.doc(hostId);

/// Provides a reference to the hosts collection for updating.
final updateHostCollectionReference =
    FirebaseFirestore.instance.collection('hosts').withConverter<UpdateHost>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a host document for updating.
DocumentReference<UpdateHost> updateHostDocumentReference({
  required String hostId,
}) =>
    updateHostCollectionReference.doc(hostId);

/// Provides a reference to the hosts collection for deleting.
final deleteHostCollectionReference =
    FirebaseFirestore.instance.collection('hosts').withConverter<DeleteHost>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a host document for deleting.
DocumentReference<DeleteHost> deleteHostDocumentReference({
  required String hostId,
}) =>
    deleteHostCollectionReference.doc(hostId);

/// Manages queries against the hosts collection.
class HostQuery {
  /// Fetches [ReadHost] documents.
  Future<List<ReadHost>> fetchDocuments({
    GetOptions? options,
    Query<ReadHost>? Function(Query<ReadHost> query)? queryBuilder,
    int Function(ReadHost lhs, ReadHost rhs)? compare,
  }) async {
    Query<ReadHost> query = readHostCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [Host] documents.
  Stream<List<ReadHost>> subscribeDocuments({
    Query<ReadHost>? Function(Query<ReadHost> query)? queryBuilder,
    int Function(ReadHost lhs, ReadHost rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadHost> query = readHostCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadHost] document.
  Future<ReadHost?> fetchDocument({
    required String hostId,
    GetOptions? options,
  }) async {
    final ds = await readHostDocumentReference(
      hostId: hostId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [Host] document.
  Stream<ReadHost?> subscribeDocument({
    required String hostId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readHostDocumentReference(
      hostId: hostId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [Host] document.
  Future<DocumentReference<CreateHost>> add({
    required CreateHost createHost,
  }) =>
      createHostCollectionReference.add(createHost);

  /// Sets a [Host] document.
  Future<void> set({
    required String hostId,
    required CreateHost createHost,
    SetOptions? options,
  }) =>
      createHostDocumentReference(
        hostId: hostId,
      ).set(createHost, options);

  /// Updates a specific [Host] document.
  Future<void> update({
    required String hostId,
    required UpdateHost updateHost,
  }) =>
      updateHostDocumentReference(
        hostId: hostId,
      ).update(updateHost.toJson());

  /// Deletes a specific [Host] document.
  Future<void> delete({
    required String hostId,
  }) =>
      deleteHostDocumentReference(
        hostId: hostId,
      ).delete();
}

File: functions/src/firestore-documents/host.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export type HostType = `farmer` | `fisherman` | `hunter` | `other`

export class ReadHost {
    constructor({
        hostId,
        path,
        imageUrl,
        displayName,
        introduction,
        hostTypes,
        urls,
        createdAt,
        updatedAt
    }: {
        hostId: string
        path: string
        imageUrl: string
        displayName: string
        introduction: string
        hostTypes: Set<HostType>
        urls: string[]
        createdAt?: Date
        updatedAt?: Date
    }) {
        this.hostId = hostId
        this.path = path
        this.imageUrl = imageUrl
        this.displayName = displayName
        this.introduction = introduction
        this.hostTypes = hostTypes
        this.urls = urls
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }

    readonly hostId: string

    readonly path: string

    readonly imageUrl: string

    readonly displayName: string

    readonly introduction: string

    readonly hostTypes: Set<HostType>

    readonly urls: string[]

    readonly createdAt?: Date

    readonly updatedAt?: Date

    private static hostTypesConverterFromJson(
        hostTypes: unknown[] | undefined
    ): Set<HostType> {
        return new Set((hostTypes ?? []).map((e) => e as HostType))
    }

    private static fromJson(json: Record<string, unknown>): ReadHost {
        return new ReadHost({
            hostId: json[`hostId`] as string,
            path: json[`path`] as string,
            imageUrl: (json[`imageUrl`] as string | undefined) ?? ``,
            displayName: (json[`displayName`] as string | undefined) ?? ``,
            introduction: (json[`introduction`] as string | undefined) ?? ``,
            hostTypes:
                json[`hostTypes`] == undefined
                    ? new Set()
                    : ReadHost.hostTypesConverterFromJson(
                          json[`hostTypes`] as unknown[]
                      ),
            urls:
                (json[`urls`] as unknown[] | undefined)?.map(
                    (e) => e as string
                ) ?? [],
            createdAt: (json[`createdAt`] as Timestamp | undefined)?.toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadHost {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadHost.fromJson({
            ...cleanedData,
            hostId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateHost {
    constructor({
        imageUrl,
        displayName,
        introduction,
        hostTypes,
        urls
    }: {
        imageUrl: string
        displayName: string
        introduction: string
        hostTypes: Set<HostType>
        urls: string[]
    }) {
        this.imageUrl = imageUrl
        this.displayName = displayName
        this.introduction = introduction
        this.hostTypes = hostTypes
        this.urls = urls
    }

    readonly imageUrl: string

    readonly displayName: string

    readonly introduction: string

    readonly hostTypes: Set<HostType>

    readonly urls: string[]

    private static hostTypesConverterToJson(
        hostTypes: Set<HostType>
    ): string[] {
        return [...hostTypes]
    }

    toJson(): Record<string, unknown> {
        return {
            imageUrl: this.imageUrl,
            displayName: this.displayName,
            introduction: this.introduction,
            hostTypes: CreateHost.hostTypesConverterToJson(this.hostTypes),
            urls: this.urls,
            createdAt: FieldValue.serverTimestamp(),
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateHost {
    constructor({
        imageUrl,
        displayName,
        introduction,
        hostTypes,
        urls,
        createdAt
    }: {
        imageUrl?: string
        displayName?: string
        introduction?: string
        hostTypes?: Set<HostType>
        urls?: string[]
        createdAt?: Date
    }) {
        this.imageUrl = imageUrl
        this.displayName = displayName
        this.introduction = introduction
        this.hostTypes = hostTypes
        this.urls = urls
        this.createdAt = createdAt
    }

    readonly imageUrl?: string

    readonly displayName?: string

    readonly introduction?: string

    readonly hostTypes?: Set<HostType>

    readonly urls?: string[]

    readonly createdAt?: Date

    private static hostTypesConverterToJson(
        hostTypes: Set<HostType>
    ): string[] {
        return [...hostTypes]
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.imageUrl != undefined) {
            json[`imageUrl`] = this.imageUrl
        }
        if (this.displayName != undefined) {
            json[`displayName`] = this.displayName
        }
        if (this.introduction != undefined) {
            json[`introduction`] = this.introduction
        }
        if (this.hostTypes != undefined) {
            json[`hostTypes`] = UpdateHost.hostTypesConverterToJson(
                this.hostTypes
            )
        }
        if (this.urls != undefined) {
            json[`urls`] = this.urls
        }
        if (this.createdAt != undefined) {
            json[`createdAt`] = this.createdAt
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteHost {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the hosts collection for reading.
 */
export const readHostCollectionReference = db
    .collection(`hosts`)
    .withConverter<ReadHost>({
        fromFirestore: (ds: DocumentSnapshot): ReadHost => {
            return ReadHost.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(`toFirestore is not implemented for ReadHost`)
        }
    })

/**
 * Provides a reference to a host document for reading.
 * @param hostId - The ID of the host document to read.
 */
export const readHostDocumentReference = ({
    hostId
}: {
    hostId: string
}): DocumentReference<ReadHost> => readHostCollectionReference.doc(hostId)

/**
 * Provides a reference to the hosts collection for creating.
 */
export const createHostCollectionReference = db
    .collection(`hosts`)
    .withConverter<CreateHost>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for CreateHost`)
        },
        toFirestore: (obj: CreateHost): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a host document for creating.
 * @param hostId - The ID of the host document to read.
 */
export const createHostDocumentReference = ({
    hostId
}: {
    hostId: string
}): DocumentReference<CreateHost> => createHostCollectionReference.doc(hostId)

/**
 * Provides a reference to the hosts collection for updating.
 */
export const updateHostCollectionReference = db
    .collection(`hosts`)
    .withConverter<UpdateHost>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for UpdateHost`)
        },
        toFirestore: (obj: UpdateHost): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a host document for updating.
 * @param hostId - The ID of the host document to read.
 */
export const updateHostDocumentReference = ({
    hostId
}: {
    hostId: string
}): DocumentReference<UpdateHost> => updateHostCollectionReference.doc(hostId)

/**
 * Provides a reference to the hosts collection for deleting.
 */
export const deleteHostCollectionReference = db
    .collection(`hosts`)
    .withConverter<DeleteHost>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for DeleteHost`)
        },
        toFirestore: (): DocumentData => {
            throw new Error(`toFirestore is not implemented for DeleteHost`)
        }
    })

/**
 * Provides a reference to a host document for deleting.
 * @param hostId - The ID of the host document to read.
 */
export const deleteHostDocumentReference = ({
    hostId
}: {
    hostId: string
}): DocumentReference<DeleteHost> => deleteHostCollectionReference.doc(hostId)

/**
 * Manages queries against the hosts collection.
 */
export class HostQuery {
    /**
     * Fetches host documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (query: Query<ReadHost>) => Query<ReadHost>
        compare?: (lhs: ReadHost, rhs: ReadHost) => number
    }): Promise<ReadHost[]> {
        let query: Query<ReadHost> = readHostCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadHost> = await query.get()
        let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadHost>) =>
            qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific host document.
     * @param hostId - The ID of the host document to fetch.
     */
    async fetchDocument({
        hostId
    }: {
        hostId: string
    }): Promise<ReadHost | undefined> {
        const ds = await readHostDocumentReference({
            hostId
        }).get()
        return ds.data()
    }

    /**
     * Adds a host document.
     * @param createHost - The host details to add.
     */
    async add({
        createHost
    }: {
        createHost: CreateHost
    }): Promise<DocumentReference<CreateHost>> {
        return createHostCollectionReference.add(createHost)
    }

    /**
     * Sets a host document.
     * @param hostId - The ID of the host document to set.
     * @param createHost - The host details to set.
     * @param options - Options for the set operation.
     */
    async set({
        hostId,
        createHost,
        options
    }: {
        hostId: string
        createHost: CreateHost
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createHostDocumentReference({
                hostId
            }).set(createHost)
        } else {
            return createHostDocumentReference({
                hostId
            }).set(createHost, options)
        }
    }

    /**
     * Updates a specific host document.
     * @param hostId - The ID of the host document to update.
     * @param updateHost - The details for updating the host.
     */
    async update({
        hostId,
        updateHost
    }: {
        hostId: string
        updateHost: UpdateHost
    }): Promise<WriteResult> {
        return updateHostDocumentReference({
            hostId
        }).update(updateHost.toJson())
    }

    /**
     * Deletes a specific host document.
     * @param hostId - The ID of the host document to delete.
     */
    async delete({ hostId }: { hostId: string }): Promise<WriteResult> {
        return deleteHostDocumentReference({
            hostId
        }).delete()
    }
}
